Deploy w docker-stack to K8s cluster

urpylka opened the pull request:
Hi! Continue our topic https://github.com/tiredpixel/postgres-xl-docker/issues/55 I've created this PR.
I've checked this and it's work. But please recheck this and let me know if something won't work.

Also I allow to edit PR by you. And you can fix my temp commit d8b78ceceb9e210bc04a34f9e1b53997a6243994 as you will want.

Thanks!

urpylka committed:
Cleanup whitespaces

urpylka committed:
Add init-eg-stack

For use it with kubernetes

urpylka committed:
Update documentation about K8S

tiredpixel commented on Dockerfile:
```diff
@@ -24,7 +24,7 @@ RUN useradd ${PG_USER} -d ${PG_HOME} && \
 WORKDIR ${PG_HOME}
 
-COPY --chown=postgres:postgres lib/ ./lib/
+COPY --chown=postgres:postgres lib/postgres-xl ./lib/postgres-xl
```
Why? I see e3dc6d6c7ab6bf95f29251eab343d57d43bb707a says it might be used for something else—except it isn't:
https://github.com/tiredpixel/postgres-xl-docker/tree/master/lib
And even if it were, would we not want all the library dependencies present in the image? How else would it be compiled?

urpylka replied:
1. I've tried to add `pgpool` as submodule. It works as separate image and it doesn't need to add to the image.

2. To be honest I hadn't seen that target directory has path `/var/lib/postgresql/lib/`. I had tried to make `pgpool` as part of the image early. But it's strange that path for `pgpool` or `wait-for-it` will consist `/var/lib/postgresql/lib/` (because it is not postgresql). Maybe we need to change it.

    For ex `postgres-xl` will be copied to `/var/lib/postgres-xl`, `wait` – `/var/lib/wait`

    It isn't need to decide instead of user. Let he own decide what he want to copy to the image and where it needs. What do think about that?

3. Why we need user in the image? Docker has the idea **single service as a container**. Or couldn't I understand your implementation?

tiredpixel replied:
Thanks for your replies.

1./2. There shouldn't be any need to add Pgpool as a submodule, unless you're compiling that rather than using the built images. But I've considered this further, including with your up-to-date replies, and I've come to the conclusion that I don't see a strong enough case for adding Pgpool to this project. The main reasons for that are that it's not actually necessary for use of Postgres-XL (indeed, I and others have been using it for a long time without), that Pgpool is by no means the only option for those wanting connection pooling (e.g. HAProxy is a popular alternative), and I don't think we can guess reasonably enough what others will need in order to provide utility in supplying a large config, without increasing complexity. Since also the Docker Stack scripts are examples only (since we also cannot predict the topology others are using, including whether there are 2 or more datanodes or coordinators), then I think it's simpler for this project to just avoid that whole question, and leave it to whomever installs the cluster however they wish (Docker Swarm, Kubernetes, etc.)

3. I think there might be some confusion, here, since I don't think I asked about why we need a user in the image, and yes of course I understand Docker's concepts. I think we can drop this point as unimportant.

urpylka replied:
1-2 yes, **I agree with you. Pg-pool is optional.** We can just write about it in the documentation (like tutorial).

What about 3 I don't still know for what we use **users politics in Docker file**.

And last I am convinced that we need to copy lib dependencies separately to the image (without any sub directory). I suggest copy `/lib/postgres-xl` to `/var/lib/postgres-xl`.

tiredpixel replied:
If we agree that Pgpool is optional, we can certainly mention it in the README—maybe along with HAProxy, etc. But in that case, there's no need for us to include it in the example configuration, and certainly not to include the 750-line example configuration file.

Sorry, I don't understand your question about 3. What user's politics? Is it possible that's a typo? I'm not aware of any politics in the Dockerfile. The user itself is because of least privilege, so it is compiled and ultimately runs as `postgres` user—similar to PostgreSQL itself (i.e. non-Postgres-XL).

Are you aware that there are *two* Docker images, here? The `Dockerfile` is used to install all the build dependencies for development, as well as to include the Postgres-XL source (it has to come from somewhere, after all). But those are not included in the final images released to Docker Hub; those are built by `ci/Dockerfile`, are are not expected to be easily runnable locally by a user (since they depend on a build system I host privately). So, it's not that the `lib/postgres-xl` files go into the final image (if you download the Docker Hub pacakged images, you'll see this); the source files are compiled, and then those compiled binaries are included in the other Docker image. Does that answer your question?

urpylka replied:
No, I mean main Dockerfile. We have politics that one process in one container why we use OS users? Why we use users inside docker?

```Dockerfile
RUN useradd ${PG_USER} -d ${PG_HOME} && \
    mkdir -p ${PG_LIB} ${PG_HOME} && \
    chown -R ${PG_USER}:${PG_USER} ${PG_LIB} ${PG_HOME}

...

USER ${PG_USER}

...

USER root

...

USER ${PG_USER}

```

What about CI I need more time to understand it.

tiredpixel replied:
Because Docker has users within the container… and dropping root privileges is standard practice from a security point of view. And this isn't an OS user; it's a user within the container.

It sounds like you might like to spend some time looking at things like the Postgres images (i.e. non-Postgres-XL—not this project), if you haven't already:
https://github.com/docker-library/postgres/blob/master/12/Dockerfile

tiredpixel commented on Dockerfile:
```diff
@@ -53,4 +53,7 @@ COPY bin/* ${PG_LIB}/bin/
 VOLUME ${PG_HOME}
+#-------------------------------------------------------------------------------
+# https://github.com/ufoscout/docker-compose-wait
+COPY wait-2.7.3 /wait
```
There's a few issues, here; more about those in other sections. But for this, the main one is there's a vendoring of a specific version which will never be upgraded by either updating libs or base images, meaning no fixes will ever be applied, non-standard usage contrary to FHS (also I'll be honest, I'm less worried about this, and I myself break that sometimes), and no checksumming or such.

tiredpixel replied:
It seems I can't add a comment to the binary file added (`wait-2.7.3`). Separate to my other points, there are some issues with this:
- I can't be sure where it's come from or about its integrity, since it's a binary blob without checksums matching anything elsewhere;
- in general, I think it's better to do a checksummed download within an image build, rather than add a binary to the repo itself, unless it's really absolutely necessary;
- it will likely never get updated, as mentioned;
- just as importantly, I'm concerned it would violate the licence; `wait` following your link lists it as Apache 2.0 licence, which has a licence and copyright notice obligation, which doesn't appear to be satisfied here. Off the top of my head, I'm also unsure about Apache 2.0-MIT compatibility; I *think* they're deemed compatible (if of course the licence requirements are satisfied), but it would be necessary to check.

urpylka replied:
Ok. I agree with you.

So as I can understand we have two problems:

1. Update & checksum - I can replace file to next strings.

    ```Dockerfile
     ADD https://github.com/ufoscout/docker-compose-wait/releases/download/2.7.3/wait /wait
     RUN chmod +x /wait
     ```
2. License - as I understood:
    First one: If we don't change wait-service, we don't need to use this license as main.
    Second: We can add information about wait-service and Apache license to Readme file.

tiredpixel replied:
I am not a lawyer, but I don't think that's necessarily the correct interpretation of the Apache licence. However, I agree it has no effect on main. Regardless, however, I've been thinking about this point further, too, and I feel it necessary to clarify: what problem, exactly, did you experience, leading to adding the wait script? Since as I mentioned, there is extensive healthchecking support in this project already, and even if containers start in the wrong order, they should usually fix themselves (after the bootstrapping process is finished, which is a specific exception for setting up the cluster). I'm currently not seeing enough justification to add the wait script, regardless of how it is installed or packaged or checksummed, since I'm unclear about what it gives us compared to what's already there.

I think this PR has become quite long, and seems to be be addressing multiple different things (e.g. Pgpool, wait script, minor fixes and extensions to documentation, additional example file for installing on Kubernetes via Docker Stack and kubectl, …). It might be easier to split this stuff apart into smaller PRs. However, I also don't want there to be wasted work, so to summarise what I understand from this PR so far:
- minor fixes and extensions to documentation are welcome (pretty much as-is);
- additional example file for installation Kubernetes via Docker Stack and kubectl also is fine, but note much of it is likely very similar to the files already in place for Docker Swarm via Docker Stack (after all, that's much of the point of the multi-orchestrator Docker Stack, providing an interface for this);
- I'm thinking that this project is likely not an appropriate place for adding Pgpool, for the reasons given above;
- I'm currently struggling to understand what issue there was with the healthchecks already present, motivating the adding of the wait script.

Hope that helps to clarify. Thanks again.

tiredpixel commented on README.md:
```diff
@@ -79,6 +79,47 @@ Note there are various caveats to using this, which you can read about in detail
+# TO UP
+docker stack deploy --orchestrator=kubernetes --namespace=default -c docker-compose.stack.yml pxl_stack
+# After you deployed pxl_stack, you need to initialize it (if it didn't do early or you purged PVC)
+./bin/init-eg-stack
```
Typically, the `init-eg-*` scripts are just that: examples (e.g.), since of course a completely different topology might be required (this image supports fewer or more than 2 coordinators or datanodes).

urpylka replied:
Yes of course, but all this solution is sample.
And to configure it for user's application all should change: `docker-compose`, `init-eg-*`.

I can add comment "for automatic configure it to test architecture, you can use it"

tiredpixel commented on README.md:
```diff
@@ -79,6 +79,47 @@ Note there are various caveats to using this, which you can read about in detail
+kubectl get pv
+
+# TO PORT FORWARD
+kubectl port-forward pg-pool-5b64b65f48-mc7g5 5432:5432
```
I'm not currently working with K8S so I can't check this, but this seems like interesting information. Thanks for being so detailed.

urpylka replied:
To be honest I forgot to anonymize this string.

tiredpixel commented on bin/docker-healthcheck-coord:
```diff
@@ -7,18 +7,18 @@ if [ ! -f "$bootstrapped" ]; then
     sleep_int=$(((RANDOM % sleep_int_max) + 1))
     echo "Sleeping to avoid race-condition (!) [$sleep_int s]"
     sleep "$sleep_int"
-    
```
Not actually accidental, but fine. I'm slowly getting used to people stripping my whitespace out even when it makes sense conceptually. ;) I'll accept.

urpylka replied:
Sometimes it really needs. For ex in Python in Pep8. But there I am not sure.

tiredpixel replied:
Yes, but this is not Python. :) Shell scripts generally aren't whitespace-sensitive in that way (there are some exceptions, such as in tabs and multiline strings), but there's no harm in them if they make sense conceptually (e.g. everything is indented). But as I say, this doesn't really matter, here.

tiredpixel commented on bin/init-eg-stack:
```diff
@@ -0,0 +1,59 @@
+#!/bin/bash -u
```
I feel like maybe there's a little confusion, here. The aim, as far as I understand it, is to support—or rather, make it easier via another method, since extensive work and testing for running on Kubernetes via Helm charts was already done by @LamaAni and @sstubbs (#35 and places)—use of this image on Kubernetes. Except most of this is titled 'stack', whereas usage of `docker stack deploy` has already been supported right from the beginning—just with the Docker Swarm orchestrator:
https://github.com/tiredpixel/postgres-xl-docker/blob/master/docker-compose.swarm.yml
https://github.com/tiredpixel/postgres-xl-docker/blob/master/docker-compose.image.yml
I'd actually expect something very similar to work already… But anyway, it's not that the focus of this is providing Stack support, but rather adding to the Stack support that's already been in place for a long time. Unless I misunderstand something…

urpylka replied:
I've not touched `docker-compose.swarm.yml`, because I don't want to test `swarm` capability.
What about  `docker-compose.image.yml` I don't still know for what it is?

Now I've seen image and swarm file is very same. Maybe we can union these, but it needs the tests.

---

What about https://github.com/tiredpixel/postgres-xl-docker/issues/35 it is very interesting. I've found information about `netcat` as `wait-for-it` in discussion. But I don't know what will be better to use. Tell me what are you thinking about that?

tiredpixel replied:
Sorry, missed this comment. `image` is for using the pre-built images published to Docker Hub, rather than compiling locally. I'm expecting that most people using this project won't rebuild themselves each time, but rather, will use the published Docker images. However, anyone wanting to develop this project, or indeed to verify the published Docker images, can of course build (like I think you're doing).

As per that discussion, `netcat` was removed and replaced by `pgxc_monitor`. But it turned out to have a use for Kubernetes, so it was readded in https://github.com/tiredpixel/postgres-xl-docker/pull/40 by @LamaAni .

tiredpixel commented on bin/init-eg-stack:
```diff
@@ -0,0 +1,59 @@
+    local node=$1
+    shift
+
+    kubectl exec "$node" -i -t -- "$@"
```
I guess it's already worth noting that this is not Kubernetes per se, but also using Kubectl, which I guess is pretty standard, but I don't *think* (perhaps someone else knows more than me about this?) that it's actually strictly necessary for running Kubernetes clusters.

urpylka replied:
Ok, I will add to documentation that `init-eg-stack` uses `kubectl`

tiredpixel commented on docker-compose.stack.yml:
```diff
@@ -0,0 +1,138 @@
+
+x-custom:
+  shared_image: &shared_image
+    image: smirart/postgres-xl:kubernetes
```
Yes, this temporary commit needs to change, please.

tiredpixel commented on docker-compose.stack.yml:
```diff
@@ -0,0 +1,138 @@
+    image: smirart/postgres-xl:kubernetes
+
+services:
+  pgpool:
```
So, Pgpool… Pgpool is excellent. But why do we need it? It's already possible to address any coordinator, and if there aren't enough spare connections, it's possible to simply add a coordinator (since all the topologies in this project, including the `init-eg-*` scripts, and the YAML files, are examples only). So this adds another piece to the stack, which I don't automatically understand is necessary. If it's for load-balancing, there are other ways of achieving that beyond Pgpool—e.g. what if someone wants to use HAProxy instead? If it's for high-availability—well, this doesn't give it, because the Pgpool instance could also fail, and there's only one of them. And if we're relying on Kubernetes to detect the health and relaunch it—isn't that what will happen for a coordinator anyway?

urpylka replied:
As I could understand when we use two or more coordinators we haven't one enter point (connection point). `pgpool` may be that enter point. Therefore I added him. But how I've agreed above, `pgpool` is optional.

tiredpixel replied:
But that still doesn't solve the issue, does it? Because unless you have multiple instances of Pgpool, there's still no redundancy there—unless you're relying on unhealthy containers being killed and relaunched. Which I think is reasonable—except that that's arguably what happens for coordinators anyway. And even if someone wanted a single entrypoint for a connection, they might be using another method entirely—such as HAProxy, or DNS failover, or a VIP. There's no reason it has to be Pgpool.

tiredpixel commented on docker-compose.stack.yml:
```diff
@@ -0,0 +1,138 @@
+      - WAIT_HOSTS=db-gtm-1:6666
+      - WAIT_BEFORE_HOSTS=20
+      - WAIT_HOSTS_TIMEOUT=60
+    command: sh -c "/wait && docker-cmd-coord"
```
In general, I'm not usually convinced by wait-for-it type scripts. My reasoning is thus: if things need a special order to start up, that can squash a lot of issues below the service which can appear cluster-wide, but *far* more importantly, it gives a false sense of security, since there's no guarantee that after a container failure or network partition everything will be relaunched in the same order. Even if you're checking health and killing off old things and relaunching the entire container, that still leads to a period where things can be running in an unplanned order cluster-wide, and also, it's very wasteful time-til-recovery-wise, since it requires a full cluster restart just to replace a dead node (e.g. a failed coordinator, especially as Postgres-XL can be so brittle, in my opinion).

This is why there's such a detailed healthcheck system already in this image, including the two-stage bootstrap process for initial initialisation of the cluster, and deep healthchecks for coordinators which even write data into a table despite the performance cost, since in my experience, Postgres-XL can fail very nastily when run in containers, and it's possible to have a coordinator which is live and even working for reads but then ultimately failing for writes. So I guess I'm not sure really what the wait-for-it gives us anyway—I don't use `depends_on` anyway, since it already isn't supported for Docker Swarm via `stack deploy`), and I'm equally concerned it provides a false sense of security. I'd be interested in hearing your reasons further about this.

urpylka replied:
To be honest I've not looked so deeper in that problem. But what I've understood: if you wanna to run async processes you need to synchronise that.

I've thought about two options:

1. To write server-client application who will check container via healthcheck script (because no one knows condition of container better than his healthcheck).
2. Yes we have third option – use orchestrator services like `depends_on` in **docker-compose**, but it hasn't been consisted everywhere (in **docker-stack** f.e.).
3. To use `wait-for-it` and meaning that developers of Postgres-XL care about thus situations (that if something goes wrong – port will be closed). **Yes, it is like take on faith, but it will be simple. Because I've selected this.**

tiredpixel replied:
1. This sounds like overkill.

2. Yes, but as noted, it's not actually necessary to start the services in this project in any particular order (excepting the bootstrapping requirement). If the GTM is started last, the cluster will converge and restart services as needed already—similar to if there is a networking blip in the night. We wouldn't want to have to wait until we're able to fix it in person, and then restart the whole cluster in a particular order. Preventing this is what the healthchecks try hard to achieve.

3. I can't take it on faith, because I know for a fact that Postgres-XL does not correctly detect all circumstances relating to unhealthy connections. :) e.g. If the network link is interrupted between a coordinator and a datanode, this often isn't even detected. Even with a select-based healthcheck, then it's possible for this to pass—and then have the coordinator fail on the first write (because the underlying links with the datanodes have been severed, or a datanode itself is unhealthy). I agree it would be nice to be able to leave this to Postgres-XL code, however—but this hasn't matched my experience of Postgres-XL issues over the last few years.

urpylka replied:
Okey, I will take one day for the test and give you response about errors related to using / not using `wait-for-it`

tiredpixel replied:
Thanks. It would be useful to gain an understanding of the specific issues you encountered leading you to add `wait-for-it`, also with any error messages. The healthchecks which are there should already handle the use-case you mention—plus much more. If there is some issue with them, those can probably be fixed without having to change the whole method.

tiredpixel commented on pgpool.conf:
```diff
@@ -0,0 +1,750 @@
+# ----------------------------
```
I'm very loathe to add a 750-line configuration file into the image unless there's a very good reason to do so. I don't even have time to read it. :)

tiredpixel commented on docker-compose.stack.yml:
```diff
@@ -0,0 +1,138 @@
+      - PG_GTM_HOST=db-gtm-1
+      - PG_GTM_PORT=6666
+      - PG_HOST=0.0.0.0
+      - PG_NODE=coord_2
```
I'd be curious to hear more about ae8378f4e956022ddd7ed920965c7e062e88f3ec , if you would be so kind. It doesn't really matter what they're called—this is why it's even exposed as a variable—but it's interesting that it's different.

urpylka replied:
I've reproduced this problem by `docker-compose.swarm.yml`

```bash
smirart@localhost postgres-xl-docker % docker stack deploy --orchestrator=kubernetes --namespace=default -c docker-compose.swarm.yml pxl_stack
top-level network "db_a" is ignored
top-level network "db_b" is ignored
service "db_data_2": network "db_a" is ignored
service "db_gtm_1": network "db_a" is ignored
service "db_coord_1": network "db_b" is ignored
service "db_coord_1": network "db_a" is ignored
service "db_coord_2": network "db_a" is ignored
service "db_coord_2": network "db_b" is ignored
service "db_data_1": network "db_a" is ignored
Stack.compose.docker.com "pxl_stack" is invalid: [spec.stack.services[0].name: Invalid value: "db_data_2": not a valid service name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[0].volumes[0]: Invalid value: "db_data_2": not a valid volume name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[1].name: Invalid value: "db_gtm_1": not a valid service name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[1].volumes[0]: Invalid value: "db_gtm_1": not a valid volume name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[2].name: Invalid value: "db_coord_1": not a valid service name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[2].volumes[0]: Invalid value: "db_coord_1": not a valid volume name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[3].name: Invalid value: "db_coord_2": not a valid service name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[3].volumes[0]: Invalid value: "db_coord_2": not a valid volume name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[4].name: Invalid value: "db_data_1": not a valid service name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.stack.services[4].volumes[0]: Invalid value: "db_data_1": not a valid volume name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]
```

Errors:

spec.stack.services[0].name: Invalid value: "db_data_2": not a valid service name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'),

spec.stack.services[0].volumes[0]: Invalid value: "db_data_2": not a valid volume name in Kubernetes: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

tiredpixel replied:
Thanks!

tiredpixel commented on docker-compose.stack.yml:
```diff
@@ -0,0 +1,138 @@
+      retries: 2
+
+volumes:
+  db-gtm-1-data: {}
```
Any reason why all these volumes are named differently to in the other files? I see c59b951a2ad88ecba106f267b78419fc3197ed82 , but I'm not really sure about the rationale… Again, it doesn't really matter what the volumes are called—but I'm curious about the inconsistency.

urpylka replied:
It's just my practice what I've used and seen in another projects. Just for me it's convenient to separate volumes and names of containers by this suffix.

tiredpixel replied:
Okay. So it's your own convention. I don't mind too much, as this is a separate file—but I note that you're using a convention from elsewhere over the one already here. But it doesn't really matter.

tiredpixel reviewed (changes requested):
That completes my first look-through. Thanks, once again! :)

p.s. By the way, have you considered just extending this image and adding whatever else you need to that (including `wait`, if you so desire)? It seems to me that it's a similar question to #53 , regarding additional locales for encoding…

urpylka committed:
Fix names

For working at K8s cluster

urpylka committed:
Cleanup

urpylka committed:
Add default healthcheck options

To understand better what is going on

urpylka commented:
I've made changes that we have been discussing here.

1. Removed `wait-for-it` because all working without him.
2. Modificated `docker-compose.swarm.yml` instead adding another one. (I've renamed him to `docker-compose.stack.yml` and fixed names for working in K8s)
3. Moved pgpool to documentation
4. Cleanup

tiredpixel reviewed (approved):
Looks much better. Thanks for your work on this!
